File: 14 Binary search algorithm/MinMaxDivision/minMaxDivision.js
function solution(K, M, A) {
  let begin = Math.max(...A);
  let end = A.reduce((acc, i) => acc + i, 0); // sum
  //let largeSum = begin + M + 1; // The large sum is the maximal sum of any block.
  let result = 0;
  while (begin <= end) {
    const mid = Math.floor((begin + end) / 2);

    let currentSum = 0;
    let blocks = 1;

    for (let i = 0; i < A.length; i++) {
      if (currentSum + A[i] > mid) {
        blocks++;
        currentSum = A[i];
        if (blocks > K) { break; }
      } else {
        currentSum += A[i];
      }
    }
    if (blocks > K) {
      begin = mid + 1;
    } else {
      result = mid;
      end = mid - 1;
    }
}
return result;
}
//console.log(solution(3, 5, [2, 1, 5, 1, 2, 2, 2]));

File: Lessons/14 Binary search algorithm/MinMaxDivision/minMaxDivision.cpp
#include <math.h>
#include <algorithm>
#include <numeric>
#include <vector>
int solution(int K, int M, vector<int> &A) {
  int begin = *std::max_element(A.begin(), A.end());
  int end = std::accumulate(A.begin(), A.end(), 0);  // sum
  int result = 0;
  while (begin <= end) {
    int mid = floor((begin + end) / 2);

    int currentSum = 0;
    int blocks = 1;

    for (int i = 0; i < A.size(); i++) {
      if (currentSum + A[i] > mid) {
        blocks++;
        currentSum = A[i];
        if (blocks > K) {
          break;
        }
      } else {
        currentSum += A[i];
      }
    }
    if (blocks > K) {
      begin = mid + 1;
    } else {
      result = mid;
      end = mid - 1;
    }
  }
  return result;
}
